docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,312 @@
+* Dependency relations between two objects are tracked using a directed
+  graph. When two objects (or nodes) are connected with a directed edge they
+  form a dependency relation.
+
+Eventhough bears still have to pass ``Result`` instances to communicate with
+coala, it is now possible to pass arbitrary python objects. This has its
+benefits in case of dependency bears since they can pass any arbitrary
+object instead of just ``Result`` objects.
+
+``DependencyBear``
+------------------
+
+Handling of bear dependencies by the old core wasn't effective. The
+old core used a queuing mechanism to communicate between bear runs and passed
+all the dependencies to the ``dependency_results`` argument of the ``run``
+function which led to problems of a ``GlobalBear`` not running if any one of its
+dependencies (e.g. a ``local bear``) terminated, the NextGen-Core improves on
+this.
+
+A new bear known as the ``DependencyBear`` is also provided by the NextGen-Core
+making it more convenient for bear developers to write dependency bears, by
+passing the dependency results using task objects. This technique of handling
+dependencies make it possible for the ``DependencyBear`` to support caching.
+The ``dependency_results`` lie inside ``self.dependecy_results`` inside a bear
+and can be accessed that way (although this is discouraged since it bypasses
+caching and could yield weird results when the core is run multiple times in a
+row).
+
+This bear serves as a base class which parallelizes tasks for each dependency
+result. A bear dependent on other bears can specify its dependencies in
+``BEAR_DEPS``. For example, there are two bears ``Foo`` and ``Bar``
+and bear ``Bar`` depends on ``Foo``. This can be written as
+
+::
+
+    class BarBear(DependencyBear):
+        BEAR_DEPS = {FooBear}
+
+This solves the dependency issues of GlobalBears on LocalBears that were there
+in the old core. Now that the new dependency management is in place GlobalBears
+won't be stalled due to the termination of a LocalBear run. This eradicates all
+the synchronization problems faced by the old core.
+
+Multiple bears can be included as a dependency of a bear in the BEAR_DEPS
+field. The results of the dependency bears are saved in a dictionary
+called ``_dependency_results`` which is initialized in the ``__init()__``
+method of the class ``Bear.py`` and can be accessed using the method
+``dependency_results()`` also belonging to the same class.
+
+Writing a DependencyBear
+------------------------
+
+Let's consider a bear to be dependent on a project bear ``Fizz`` and a file bear
+``Buzz`` then the corresponding DependencyBear let's call it ``FizzBuzz`` will
+look like the following:
+
+::
+
+    class FizzBuzzBear(DependencyBear):
+        BEAR_DEPS = {FizzBear, BuzzBear}
+
+        def analyze(self, dependency_bear, dependency_result, a_number=100):
+            yield '{} ({}) - {}'.format(
+                dependency_bear.name, a_number, dependency_result)
```

Comment:
Also implement `Fizz` and `Buzz` bears here :)
And write what results we would receive :)